README.md
[![Travis-CI Build Status](https://travis-ci.org/jefferis/rcatmaid.svg?branch=master)](https://travis-ci.org/jefferis/rcatmaid) [![DOI](https://zenodo.org/badge/25650381.svg)](https://zenodo.org/badge/latestdoi/25650381)
# catmaid

This package provides access to the [CATMAID](http://catmaid.org/) API for 
[R](http://r-project.org/) users.  At present it provides low level functions 
for appropriately authenticated GET/POST requests, optionally parsing JSON responses.
There are also intermediate level functions that retrieve skeleton (i.e. neuron) 
information, connectivity information for one or more neurons as well as a number 
of other API endpoints. Finally, there is
a high level function to convert neurons to the representation of the
[nat](https://github.com/jefferis/nat) (NeuroAnatomy Toolbox) R package, enabling
a wide variety of analyses.

It is strongly recommended to read through the instructions below, the [package
overview documentation](http://jefferis.github.io/rcatmaid/reference/catmaid-package.html)
and then skim the [reference documentation index](http://jefferis.github.io/rcatmaid/reference/)
, which groups the available functions into useful categories.

## Quick start
```r
# install
if (!require("devtools")) install.packages("devtools")
# nb repo is rcatmaid, but R package name is catmaid
devtools::install_github("jefferis/rcatmaid")

# use 
library(catmaid)

# general help starting point
?catmaid

# examples
example(catmaid_login)
example(catmaid_fetch)
example(catmaid_get_compact_skeleton)
example(catmaid_get_neuronnames)

# use with nat
library(nat)
nl=read.neurons.catmaid(c(10418394,4453485), pid=1)
open3d()
# nb this also plots the connectors (i.e. synapses) 
# red = presynapses, cyan = postsynapses
plot3d(nl, WithConnectors=TRUE)
```
## Fancier example
This produces a 3D plot of the first and second order olfactory neurons
coloured according to the peripheral odorant receptor.
```r
# fetch olfactory receptor neurons
orns=read.neurons.catmaid("name:ORN (left|right)", .progress='text')
# calculate some useful metadata
orns[,'Or']= factor(sub(" ORN.*", "", orns[,'name']))

# repeat for their PN partners, note use of search by annotation
pns=read.neurons.catmaid("annotation:ORN PNs$", .progress='text')
pns[,'Or']= factor(sub(" PN.*", "", pns[,'name']))
# plot, colouring by odorant receptor
plot3d(orns, col=Or)
# note that we plot somata with a radius of 1500 nm
plot3d(pns, col=Or, soma=1500)
```
## Even fancier example
This follows on from the previous example. It identifies downstream partner
neurons of the ORNs and plots them in 3d coloured by their synaptic strength.
It then carries out morphological clustering with [NBLAST](http://bit.ly/nblast)
and plots the partner neurons according to those clusters.

```r
# find all the ORN downstream partners with at least 2 synapses
orn_partners=catmaid_query_connected(orns[,'skid'], minimum_synapses = 2)
# keep the ones not already in our set of PNs
# there are lots!
non_pn_downstream_ids=setdiff(unique(orn_partners$outgoing$partner), pns[,'skid'])
# download and plot those neurons
non_pn_downstream=read.neurons.catmaid(non_pn_downstream_ids, .progress='text')
plot3d(non_pn_downstream, col='grey', soma=1000)

# remove the last set of plotted neurons
npop3d()

## Plot, but colouring partners by number of synapses they receive from ORNs
# first collect those synapse numbers
library(dplyr)
totsyndf=orn_partners$outgoing %>% 
  group_by(partner) %>% 
  summarise(totsyn=sum(syn.count)) %>% 
  arrange(desc(totsyn))
hist(totsyndf$totsyn)
# now do the plot
clear3d()
# matlab style palette
jet.colors <-
colorRampPalette(c("#00007F", "blue", "#007FFF", "cyan",
"#7FFF7F", "yellow", "#FF7F00", "red", "#7F0000"))
# plot colouring by synapse number on a log scale 
# note that it is necessary to convert totsyndf$partner to a character
# vector to ensure that they are not treated as integer indices
plot3d(as.character(totsyndf$partner),  db=c(pns, non_pn_downstream), 
  col=jet.colors(10)[cut(totsyndf$totsyn, breaks = 2^(0:10))], soma=1000)

# Now let's cluster these other connected neurons
library(nat.nblast)
# convert to nblast-compatible format
# nb also convert from nm to um, resample to 1µm spacing and use k=5
# nearest neighbours of each point to define tangent vector
non_pn_downstream.dps=dotprops(non_pn_downstream/1e3, k=5, resample=1, .progress='text')
# now compute all x all NBLAST scores and cluster
non_pn_downstream.aba=nblast_allbyall(non_pn_downstream.dps, .progress='text')
non_pn_downstream.hc=nhclust(scoremat = non_pn_downstream.aba)
# plot result of clusterting as dendrogram, labelled by neuron name (rather than id)
plot(non_pn_downstream.hc, label=non_pn_downstream[,'name'])
# open new window
nopen3d()
# plot in 3d cutting into 2 clusters essentially left right
plot3d(non_pn_downstream.hc,db=non_pn_downstream, k=2, soma=1000)
clear3d() 
# 4 clusters - note local and projection neurons, gustatory neurons
plot3d(non_pn_downstream.hc,db=non_pn_downstream, k=4, soma=1000)
```

## Authentication
You will obviously need to have the login details of a valid CATMAID instance to try 
this out. As of December 2015 CATMAID is moving to token based authentication. For this
you will need to get an API token when you are logged into the CATMAID web 
client in your browser. See http://catmaid.github.io/dev/api.html#api-token for
details. 

Once you have the login information you can use the `catmaid_login` function to 
authenticate. The minimal information is your server URL and your CATMAID token.

```r
catmaid_login(server="https://mycatmaidserver.org/catmaidroot",
              authname="Calvin",authpassword="hobbes",
              token="9944b09199c62bcf9418ad846dd0e4bbdfc6ee4b")
```

Note that the CATMAID servers that I am aware of require two layers of password
protection, an outer HTTP auth type user/password combination as well as an inner
CATMAID-specific token based login. The outer HTTP auth type user/password 
combination may be specific to you or generic to the project.

### Setting environment variables in your .Renviron file
It is recommended that you set your login details by including code like 
this in in your [.Renviron file](https://www.rdocumentation.org/packages/base/versions/3.4.0/topics/Startup):

```r
catmaid.server="https://mycatmaidserver.org/catmaidroot"
catmaid.token="9944b09199c62bcf9418ad846dd0e4bbdfc6ee4b"

# additional security for mycatmaidserver.org/catmaidroot page
catmaid.authname="Calvin"
catmaid.authpassword="hobbes"
```
Be sure to leave one blank line at the end of the .Renviron file, or it will not work.

In this way authentication will happen transparently as required by all functions
that interact with the specified CATMAID server. 

### Setting environment variables in your .Rprofile file
Alternatively you can set package options in your .Rprofile file, but the 
environment variable approach is now recommended as it handles a few edge cases
where options are not read by R processes e.g. when building vignettes.

```r
options(catmaid.server="https://mycatmaidserver.org/catmaidroot",
  catmaid.authname="Calvin",catmaid.authpassword="hobbes",
  catmaid.token="9944b09199c62bcf9418ad846dd0e4bbdfc6ee4b")
```
Once again, be sure to leave one blank line at the end of the .Rprofile file, 
or it will not work.

### Cached authentication 
Whether you use options in your `.Renviron` as described above or you login 
explicitly at the start of a session using `catmaid_login` the access credentials 
will be cached for the rest of the session. You can still authenticate explicitly
to a different CATMAID server (using `catmaid_login`) if you wish.

### Multiple servers
If you use more than one CATMAID server but always do so in different sessions
or rmarkdown scripts then you can save an appropriate `.Renviron` file in the 
project folder.

If you need to talk to more than one CATMAID server in a single session then you 
must use `catmaid_login` to login into each server

```r
# log in to default server specified in .Renviron/.Rprofile
conn1=catmaid_login()
# log into another server, presumably with different credentials
conn2=catmaid_login(server='https://my.otherserver.com', ...)
```

and then use the returned connection objects with any calls you make e.g.

```r
# fetch neuron from server 1
n1=read.neuron(123, conn=conn1)
# fetch neuron from server 2
n2=read.neuron(123, conn=conn2)
```
n.b. you must use connection objects to talk to both servers because if no 
connection object is specified, the last connection will be re-used.

## Installation
Currently there isn't a released version on [CRAN](https://cran.r-project.org/)
but can use the **devtools** package to install the development version:

```r
if (!require("devtools")) install.packages("devtools")
devtools::install_github("jefferis/rcatmaid")
```

Note: Windows users need [Rtools](https://cran.r-project.org/bin/windows/Rtools/) and
[devtools](https://cran.r-project.org/package=devtools) to install this way.

## Acknowledgements

Based on python code presently visible at:

* https://github.com/catmaid/CATMAID/blob/master/scripts/remote/access.py
* https://github.com/catmaid/CATMAID/blob/master/django/applications/catmaid/urls.py
* https://github.com/schlegelp/CATMAID-to-Blender/blob/master/CATMAIDImport.py

by Albert Cardona and Philipp Schlegel. Released under the GPL-3 license.


NEWS
catmaid 0.9.0
=============
This version is synced with the upcoming release of nat v1.9.0 which streamlines
some functionality but has some breaking changes in behaviour. It must be used
with nat >=1.8.12.900 from github.

* switch to Ops.catmaidneuron to match nat 1.9.0 in prep (and add version 
  requirement)
* Simple search for skids with exact annotation match (#89)
  Now you can do: catmaid_skids("ORN") or read.neurons.catmaid("ORN")
* Fix bug in catmaid_skids error message when looking for one skid (#90)

catmaid 0.8.1
=============
* Add function to get volumes from catmaid (#82)
* Add functions to read/write catmaid json selection files
* make catmaid_neuronnames returns NA for NA skids

catmaid 0.8
===========
This is a major version update because of the new support (and recommendation)
to use environment variables for login (.Rprofile is still supported).

* support fetching login informaion from environment variables (#80)
* catmaid_get_compact_skeleton now returns list of vectors (#79)
  (used to be a list of lists, but this was less convenient and not helpful)
* fix query_by_neuron_or_annotation adding empty annotations attribute

catmaid 0.7.2
=============
* Add catmaid_get_label_stats (use it to find neurons with soma)
* Add copy_tags_connectors (bring over tags/connector info)
* improve summary.catmaidneuron (#78)
* plot3d bugfix when there are NA points (and soma=T) (#77)

catmaid 0.7.1
=============
* only plot soma when explicitly tagged  (rather than plotting a sphere at the 
  root of the neuron if not tagged) (#62)
* only plot3d connectors if they exist (#73)
* Ensure nodes do not normally plot for catmaidneurons  (#74)
* doc fix for catmaid_get_treenode_table (#75)

catmaid 0.7
===========
This is a substantial release with many new functions/enhancements, changes to
match the API defined by the 2016.10.18 CATMAID release, support for token-based
authentication and numerous bug fixes.

* support for CATMAID token-based authentication (#35)
* Add catmaid_get_connectors_between to return connections between (multiple) 
  pre and postsynaptic partners - thanks to Zhihao Zheng (#53, #56)
* Add function to parse CATMAID urls e.g. to extract xyz position (#55)
* Add rename_neuron function (#54)
* Add catmaid_version function (#58)
* Add catmaid_user_history function (#57)
* Add catmaid_get_annotations_for_skeletons function (#38)
* Add catmaid_remove_annotations_for_skeletons function (#43)
* Add catmaid_add_volume to add 3D meshes to catmaid (#69)

* catmaid_get_connector_table can now accept multi skid specifications (#49)
* catmaid_get_connectors_between returns more informative errors (#51)

* fix "No encoding supplied: defaulting to UTF-8." warning (#59)
* stop catmaid_skids returning duplicates when there are multiple matching 
  annotations (#44)
* ensure that connectors are also scaled when xforming neurons (#47)
* catmaid_query_by_annotation returns neuron column as character vector (#46)
* catmaid_get_connector_table returns partner_skid columns as an integer (#48)
* fix bug in url for catmaid_get_review_status function (#45)
* Give sensible name to 1st column of connector data.frame (#42)
* ensure neurons without connectors scale properly (#41)
* fix bug in catmaid_connection_setenv() when some elements are null (#50)
* catmaid_get_annotations_for_skeletons - fix bug in skid order (#39)
* ensure catmaid_get_contributor_stats passes on connection (#37)
* ensure all functions pass on connection to catmaid_fetch (#60)
* connectors.neuronlist should handle neurons without connectors (#29)
* catmaid_skids should pass on pid (#64)
* fix catmaid_get_connector_table to reflect 2016.09.01-65 API change (#65)
* fix bug in catmaid_user_history with small queries (#70)

catmaid 0.6
===========
* support for upcoming httr 1.0 release
* fix: ensure catmaid connection objects are passed to all downstream functions
  e.g. so that you can you use different servers in the same session.
* fix queries that match multiple annotations when using a defined connection
  object
* fix: ensure that catmaid_fetch works even if server does not have a terminal /
  and remote path is also missing a slash
* fix: servers don't have to be https

catmaid 0.5
===========
This release 
* check for http status errors in catmaid (in case there is trouble connecting
  to site, bad URL etc)
* ... and errors in returned JSON when request is invalid
* export funcs to get/set login details as environment variables
* ... and clear those env vars (all useful for testing)

catmaid 0.4
===========
This release significantly enhances functionality for querying for neurons by
annotation or name, while also simplifying a number of plotting/analysis tasks
involving connectors (synapses).

* catmaid_query_by_neuronname was renamed to catmaid_query_by_name (since it can
  query by both neuron or annotation name as well as returning both annotations 
  and neurons)
* read.catmaid.neurons has richer attached metadata 
  (see https://github.com/jefferis/rcatmaid/issues/9)
* read.catmaid.neurons can make simple queries 
  (see https://github.com/jefferis/rcatmaid/issues/9)
* add catmaid_query_by_annotation to fetch objects tagged with a matching 
  annotation (rather than searching by object name).
* add connectors to get connector (synapse) information from a neuron
* add catmaid_catmaid_get_connector_table to see a list of incoming/outgoing
  connections for a given neuron.
* add plot3d.catmaidneuron to enable convenient plotting of synapses
* fix catmaid_get_neuronnames should return names in the same order as passed 
  skids (see https://github.com/jefferis/rcatmaid/issues/6)
* the catmaid skeleton_id is consistently referred to as skid in argument names
  and return values
* depend on nat (and therefore rgl). Import rather than depend on jsonlite.
* minor doc fixes

catmaid 0.3
===========
* cache login credentials for automatic reuse
* option to set/get environment variables for passwords etc

R/catmaid_nat.R
#' Read neuron(s) from CATMAID server into NeuroAnatomy Toolbox (nat) format
#' 
#' \code{read.neuron.catmaid} reads a single neuron, while 
#' \code{read.neurons.catmaid} generates a \code{\link[nat]{neuronlist}} object 
#' including some metadata information.
#' 
#' @details These functions provide a bridge between CATMAID and the 
#'   neuronanatomy toolbox R package (\url{https://github.com/jefferis/nat}), 
#'   which provides extensive functionality for analysing and plotting neurons 
#'   within the context of temaplate brains.
#'   
#'   Note that the soma is set by inspecting CATMAID tags that 
#'   (case-insensitively) match the regex \code{"(cell body|soma)"}. Where >1 
#'   tag exists the one that tags an endpoint is preferred.
#' @param skid A numeric skeleton id
#' @param pid Project id (default 1)
#' @param conn A \code{catmaid_connection} objection returned by 
#'   \code{\link{catmaid_login}}. If \code{NULL} (the default) a new connection 
#'   object will be generated using the values of the \bold{catmaid.*} package 
#'   options as described in the help for \code{\link{catmaid_login}}.
#' @param ... Additional arguments passed to the \code{\link{catmaid_fetch}} 
#'   function
#' @return a \code{\link[nat]{neuron}} or \code{\link[nat]{neuronlist}} object 
#'   containing one or more neurons. These neurons will have an additional class
#'   \code{catmaidneuron} which provides for some extra functionality in certain
#'   methods.
#' @seealso \code{\link{plot3d.catmaidneuron}}, \code{\link[nat]{read.neuron}}, 
#'   \code{\link{connectors}} to extract connector information from a 
#'   \code{catmaid.neuron}
#' @export
read.neuron.catmaid<-function(skid, pid=1L, conn=NULL, ...) {
  res=catmaid_get_compact_skeleton(pid=pid, skid=skid, conn=conn, ...)
  if(!length(res$nodes)) stop("no valid nodes for skid:", skid)
  swc=with(res$nodes, 
           data.frame(PointNo=id, Label=0, X=x, Y=y, Z=z, W=radius*2, Parent=parent_id)
  )
  swc$Parent[is.na(swc$Parent)]=-1L
  sp=somapos.catmaidneuron(swc=swc, tags=res$tags)
  soma_id_in_neuron = if(nrow(sp)==0) NULL else sp$PointNo
  n=nat::as.neuron(swc, origin=soma_id_in_neuron, skid=skid)
  
  # add all fields from input list except for nodes themselves
  n[names(res[-1])]=res[-1]
  # we expect connectors field to be null when there are no connectors
  if(length(n$connectors)<1) n$connectors=NULL
  fields_to_include=c("url", "headers")
  n[fields_to_include]=attributes(res)[fields_to_include]
  class(n)=c('catmaidneuron', 'neuron')
  n
}

somapos.catmaidneuron <- function(x, swc=x$d, tags=x$tags, skid=NULL, ...) {
  # Find soma position, based on plausible tags
  soma_tags<-grep("(cell body|soma)", ignore.case = T, names(tags), value = T)
  # soma is the preferred tag - use this for preference if it exists
  if(any(soma_tags=="soma")) soma_tags="soma"
  soma_id=unlist(unique(tags[soma_tags]))
  soma_id_in_neuron=intersect(soma_id, swc$PointNo)
  
  soma_d=swc[match(soma_id_in_neuron,swc$PointNo),,drop=FALSE]
  if(length(soma_id_in_neuron)>1) {
    if(sum(soma_d$Parent<0) == 1 ) {
      # just one end point is tagged as soma, so go with that
      soma_d[soma_d$Parent<0,, drop=FALSE]
    } else {
      warning("Ambiguous points tagged as soma in neuron: ",skid,". Using first")
      soma_d[1,, drop=FALSE]
    }
  } else soma_d
}

#' @rdname read.neuron.catmaid
#' @param skids One or more numeric skeleton ids or a character vector defining
#'   a query (see \code{\link{catmaid_skids}} or examples for the syntax).
#' @param OmitFailures Whether to omit neurons for which \code{FUN} gives an 
#'   error. The default value (\code{NA}) will result in nlapply stopping with 
#'   an error message the moment there is an eror. For other values, see 
#'   details.
#' @param df Optional data frame containing information about each neuron
#'   
#' @details When \code{OmitFailures} is not \code{NA}, \code{FUN} will be 
#'   wrapped in a call to \code{try} to ensure that failure for any single 
#'   neuron does not abort the nlapply/nmapply call. When 
#'   \code{OmitFailures=TRUE} the resultant neuronlist will be subsetted down to
#'   return values for which \code{FUN} evaluated successfully. When 
#'   \code{OmitFailures=FALSE}, "try-error" objects will be left in place. In 
#'   either of the last 2 cases error messages will not be printed because the 
#'   call is wrapped as \code{try(expr, silent=TRUE)}.
#'   
#'   The optional dataframe (\code{df}) detailing each neuron should have 
#'   \code{rownames} that match the names of each neuron. It would also make 
#'   sense if the same key was present in a column of the data frame. If the 
#'   dataframe contains more rows than neurons, the superfluous rows are dropped
#'   with a warning. If the dataframe is missing rows for some neurons an error 
#'   is generated. If SortOnUpdate is TRUE then updating an existing neuronlist 
#'   should result in a new neuronlist with ordering identical to reading all 
#'   neurons from scratch.
#' @export
#' @seealso \code{\link{catmaid_skids}}
#' @examples
#' \dontrun{
#' library(nat)
#' nl=read.neurons.catmaid(c(10418394,4453485))
#' plot3d(nl)
#' 
#' ## Full worked example looking at Olfactory Receptor Neurons
#' # read in ORNs (using exact match to ORN annotation)
#' # note that use a progress bar drop any failures
#' orns=read.neurons.catmaid("ORN", OmitFailures = T, .progress='text')
#' 
#' # Add two extra columns to the attached data.frame
#' # for the Odorant receptor genes and the side of brain
#' orns[,'side']=factor(ifelse(grepl("left", orns[,'name']), "L", "R"))
#' orns[,'Or']= factor(sub(" ORN.*", "", orns[,'name']))
#'    
#' # check what we have
#' # see ?head.neuronlist and ?with.neuronlist for details of how this works
#' head(orns)
#' with(orns, ftable(side~Or))
#' 
#' # now some plots
#' open3d()
#' # colour by side of brain
#' plot3d(orns, col=side)
#' clear3d()
#' # colour by Odorant Receptor
#' # note similar position of axon terminals for same ORN class on left and right
#' plot3d(orns, col=Or)
#' 
#' ## Additional example using Olfactory Projection Neurons
#' pns=read.neurons.catmaid("annotation:ORN PNs$", .progress='text')
#' pns[,'side']=factor(ifelse(grepl("left", pns[,'name']), "L", "R"))
#' pns[,'Or']= factor(sub(" PN.*", "", pns[,'name']))
#' 
#' # check that we have the same levels for the Odorant Receptor (Or) factor
#' # for ORNs and PNs
#' stopifnot(levels(pns[,'Or'])==levels(orns[,'Or']))
#' 
#' # Ok, let's plot the PNs - they will be in matching colours
#' plot3d(pns, col=Or)
#' 
#' }
read.neurons.catmaid<-function(skids, pid=1L, conn=NULL, OmitFailures=NA, df=NULL, ... ) {
  skids=catmaid_skids(skids, conn = conn, pid=pid)
  if(is.null(df)) {
    names(skids)=as.character(skids)
    df=data.frame(pid=pid, skid=skids, 
                  name=catmaid_get_neuronnames(skids, pid, conn=conn),
                  stringsAsFactors = F)
    rownames(df)=names(skids)
  } else {
    names(skids)=rownames(df)
  }
  fakenl=nat::as.neuronlist(as.list(skids), df=df)
  nat::nlapply(fakenl, read.neuron.catmaid, pid=pid, conn=conn, OmitFailures=OmitFailures, ...)
}

#' Get data.frame of connector (synapse) information from a neuron or neuronlist
#' 
#' In contrast to \code{\link{catmaid_get_connector_table}} this assumes that
#' you have already read the neurons into an R structure of class 
#' \code{\link[nat]{neuron}} or \code{\link[nat]{neuronlist}}.
#' 
#' @param x Neuron or neuronlist
#' @param ... Additional arguments passed to methods (and to \code{nlapply} in 
#'   the case of connectors)
#' @return A data.frame with columns \itemize{
#'   
#'   \item treenode_id (NB this is the treenode id for the \emph{current} skeleton)
#'   
#'   \item connector_skid
#'   
#'   \item prepost integer indicating whether connection is pre-(\code{0}) or 
#'   post(\code{1})-synaptic with respect to the current neuron. In other words 
#'   this field will be \code{0} (pre) for the output synapses of this neuron.
#'   
#'   \item x Spatial Location
#'   
#'   \item y
#'   
#'   \item z
#'   
#'   \item skid For \code{connectors.neuronlist}, the skid of the skeleton from
#'   which connector information was retrieved.
#'   
#'   }
#'   
#' @export
#' @examples
#' \dontrun{
#' ornsl=read.neurons.catmaid("name:ORN left", OmitFailures = T, .progress='text')
#' conndf=connectors(ornsl)
#' summary(connectors(ornsl))
#' 
#' # plot points in 3d
#' library(nat)
#' nopen3d()
#' points3d(xyzmatrix(conndf), col=c(pre='red', post='cyan')[conndf$prepost+1])
#' }
#' @family connectors
connectors<-function(x, ...) UseMethod('connectors')

#' @rdname connectors
#' @export
connectors.catmaidneuron<-function(x, ...) {
  x[['connectors']]
}

connectors.neuron<-function(x, ...) {
  stop("This neuron does not have class 'catmaid.neuron' and therefore does",
       " not have connector information!")
}

#' @rdname connectors
#' @param subset, optional subset of neurons to keep (see 
#'   \code{\link[nat]{nlapply}} for details)
#' @export
#' @seealso \code{\link[nat]{nlapply}}
#' @importFrom plyr rbind.fill
connectors.neuronlist<-function(x, subset=NULL, ...) {
  dfs=nat::nlapply(x, FUN=connectors, ..., subset=subset)
  # drop any null return values
  dfs=dfs[!sapply(dfs,is.null)]
  df=rbind.fill(dfs)
  df$skid=as.integer(rep(names(dfs), sapply(dfs, nrow)))
  df
}

#' Plot skeleton and connectors for neuron retrieved from CATMAID
#' 
#' @export
#' @method plot3d catmaidneuron
#' @param x A neuron to plot
#' @param WithConnectors logical indicating whether or not to plot connectors
#'   (defaults to \code{FALSE}).
#' @param WithNodes logical indicating whether to plot branch/end points 
#'   (defaults to \code{FALSE} since they will obscure the synapses).
#' @param ... additional arguments passed to \code{\link[nat]{plot3d.neuron}} 
#'   (synapses). Default: \code{TRUE}.
#' @seealso \code{\link[nat]{plot3d.neuron}}
#' @importFrom rgl plot3d points3d
#' @inheritParams nat::plot3d.neuron
#' @examples 
#' \dontrun{
#' nl=read.neurons.catmaid(c(10418394,4453485))
#' plot3d(nl)
#' 
#' # now with connectors (i.e. synapses)
#' plot3d(nl, WithConnectors=TRUE)
#' }
#' @aliases plot3d
plot3d.catmaidneuron<-function(x, WithConnectors=FALSE, WithNodes=FALSE, soma=FALSE, ...) {
  soma=plot3d_somarad(x, soma)
  rglreturnlist=NextMethod(WithNodes=WithNodes, soma=soma)
  if (WithConnectors) {
    conndf = connectors(x)
    # only try to plot if the neuron has connectors
    if (length(conndf)){
      rglreturnlist[['synapses']] = 
        points3d(xyzmatrix(conndf), col = c(pre ='red', post = 'cyan')[conndf$prepost + 1])
    }
  }
  invisible(rglreturnlist)
}

# private function to choose plotting radius for a neuron
plot3d_somarad <- function(x, soma=FALSE){
  if(is.logical(soma) && !soma) {
    # do nothing, soma is FALSE
  } else if(is.numeric(soma) && soma<=0){
    # this signals that we really want to plot a sphere at the origin
    # regardless of whether it is tagged as a soma or not
    soma=-soma
  } else {
    # check if this neuron has a soma and whether it has a sensible radius
    sp=somapos.catmaidneuron(x)
    # no soma defined so we don't want to plot
    if(nrow(sp)==0) soma=FALSE
    else if(is.numeric(soma) && is.finite(soma)) {
      # if we got passed a numeric value for soma then accept that
    } else {
      # we didn't get passed a numeric value for soma, so let's 
      # see if we can come up with a sensible number
      if(sp[,'W']<=0) soma=TRUE else {
        soma=sp[,'W']/2
        # check that we don't have a stupid value e.g. because diameter
        # was not transformed
        if(soma>max(x$d[,c("X","Y","Z")], na.rm = TRUE))
          soma=T
      }
    }
  }
  soma
}

#' @export
`Ops.catmaidneuron` <- function(e1, e2) {
  n2=NextMethod()
  conndf=connectors(e1)
  if(!is.null(conndf)) {
    # multiply connectors as well
    lx=length(e2)
    if(lx==1) xyzmatrix(conndf)=xyzmatrix(conndf)*e2
    else {
      xyzmatrix(conndf)=t(t(xyzmatrix(conndf)*e2[1:3]))
    }
    n2[['connectors']]=conndf
  }
  n2
}


nsoma <- function(x) length(x[['tags']][['soma']])

#' @export
summary.catmaidneuron<-function(object, ...) {
  df=NextMethod()
  nc=nrow(connectors(object))
  if(is.null(nc)) nc=0L
  df$connectors=nc
  df$nsoma=nsoma(object)
  df
}

#' Copy the tag / connector info from a catmaid neuron to a new neuron
#' @description This function is intended primarily for developer use as a way 
#'   to copy over connector/tag information typically contained in CATMAID
#'   neurons but lost when nat functions are applied to transform a neuron.
#' @details Both connectors and tag information contain node ids. These may need
#'   to be mapped onto new node ids if these are not consistent between old and 
#'   new neurons.
#' @export
#' @param new,old information will be copied from old -> new
#' @param update_node_ids whether to update the node ids (default \code{TRUE})
#' @return A new neuron containing tag and connector information from the old 
#'   neuron and having the same class as the old neuron.
#' @family catmaidneuron
#' @importFrom nabor knn
copy_tags_connectors <- function(new, old, update_node_ids=TRUE) {
  ## connectors
  c = connectors(old)
  if(update_node_ids) {
  nnres = nabor::knn(
    data = nat::xyzmatrix(new),
    query = nat::xyzmatrix(c),
    k = 1
  )
  # note that we map indices into the point array onto PointNo
  c$treenode_id=new$d$PointNo[nnres$nn.idx]
  }
  new[['connectors']] = c

  ## tags
  # replace the tag ids using a similar strategy
  old_tag_ids = unlist(old$tags, use.names = F)
  if (!update_node_ids) {
    new[['tags']] = old[['tags']]
  } else {
    nnres = nabor::knn(
      data = nat::xyzmatrix(new),
      query = nat::xyzmatrix(old)[match(old_tag_ids, old$d$PointNo), ],
      k = 1
    )
    new_tag_ids = new$d$PointNo[nnres$nn.idx]
    new$tags = old$tags
    k = 0
    for (i in seq_along(old$tags)) {
      for (j in seq_along(old$tags[[i]])) {
        k = k + 1
        new$tags[[i]][[j]] = new_tag_ids[k]
      }
      names(new$tags[[i]]) = names(old$tags[[i]])
    }
    names(new$tags) = names(old$tags)
  }
  class(new)=class(old)
  new
}
